# Análise de dados estruturados

In [1]:
import pandas as pd

![panda](https://media.giphy.com/media/HDR31jsQUPqQo/giphy.gif)

## Criar um dataframe a partir de um dicionário

In [2]:
dict = {"country": ["Brazil", "Russia", "India", "China", "South Africa"],
       "capital": ["Brasilia", "Moscow", "New Dehli", "Beijing", "Pretoria"],
       "area": [8.516, 17.10, 3.286, 9.597, 1.221],
       "population": [200.4, 143.5, 1252, 1357, 52.98] }

In [3]:
# Transformar o dicionário em um dataframe
brics = pd.DataFrame(dict)

In [4]:
# Ver os primeiros registros desse dataframe
brics.head()

country    capital    area  population
0        Brazil   Brasilia   8.516      200.40
1        Russia     Moscow  17.100      143.50
2         India  New Dehli   3.286     1252.00
3         China    Beijing   9.597     1357.00
4  South Africa   Pretoria   1.221       52.98

## Importar um csv com o pandas

Vamos utilizar os dados que o Kaggle lançou no ano de 2017 sobre Cientistas de Dados e Data Science. São 5 datasets diferentes:

 - **schema.csv**: a CSV file with survey schema. This schema includes the questions that correspond to each column name in both the multipleChoiceResponses.csv and freeformResponses.csv.
 - **multipleChoiceResponses.csv**: Respondents' answers to multiple choice and ranking questions. These are non-randomized and thus a single row does correspond to all of a single user's answers. 
 -**freeformResponses.csv:** Respondents' freeform answers to Kaggle's survey questions. These responses are randomized within a column, so that reading across a single row does not give a single user's answers.
 - **conversionRates.csv**: Currency conversion rates (to USD) as accessed from the R package "quantmod" on September 14, 2017
 - **RespondentTypeREADME.txt**: This is a schema for decoding the responses in the "Asked" column of the schema.csv file.

In [5]:
# Carregue o dataset multipleChoiceResponses com o pandas 
multiple_choice = pd.read_csv('/home/cintia/workspace/tera-datascience-out2018/07-structured-data-analysis/multipleChoiceResponses.csv', encoding='ISO-8859-1')

/home/cintia/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (31,83,86,87,98,99,109,116,123,124,127,129,130,164) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [6]:
# Veja as primeiras linhas do dataset
multiple_choice.head()

GenderSelect        Country   Age  \
0  Non-binary, genderqueer, or gender non-conforming            NaN   NaN   
1                                             Female  United States  30.0   
2                                               Male         Canada  28.0   
3                                               Male  United States  56.0   
4                                               Male         Taiwan  38.0   

                                    EmploymentStatus StudentStatus  \
0                                 Employed full-time           NaN   
1                 Not employed, but looking for work           NaN   
2                 Not employed, but looking for work           NaN   
3  Independent contractor, freelancer, or self-em...           NaN   
4                                 Employed full-time           NaN   

  LearningDataScience CodeWriter CareerSwitcher  \
0                 NaN        Yes            NaN   
1                 NaN        NaN            NaN   
2                 NaN        NaN            NaN   
3                 NaN        Yes            NaN   
4                 NaN        Yes            NaN   

              CurrentJobTitleSelect TitleFit              ...                \
0             DBA/Database Engineer     Fine              ...                 
1                               NaN      NaN              ...                 
2                               NaN      NaN              ...                 
3  Operations Research Practitioner   Poorly              ...                 
4                Computer Scientist     Fine              ...                 

  JobFactorExperienceLevel JobFactorDepartment  JobFactorTitle  \
0                      NaN                 NaN             NaN   
1                      NaN                 NaN             NaN   
2           Very Important      Very Important  Very Important   
3                      NaN                 NaN             NaN   
4                      NaN                 NaN             NaN   

  JobFactorCompanyFunding JobFactorImpact JobFactorRemote JobFactorIndustry  \
0                     NaN             NaN             NaN               NaN   
1                     NaN             NaN             NaN               NaN   
2          Very Important  Very Important  Very Important    Very Important   
3                     NaN             NaN             NaN               NaN   
4                     NaN             NaN             NaN               NaN   

  JobFactorLeaderReputation JobFactorDiversity JobFactorPublishingOpportunity  
0                       NaN                NaN                            NaN  
1        Somewhat important                NaN                            NaN  
2            Very Important     Very Important                 Very Important  
3                       NaN                NaN                            NaN  
4                       NaN                NaN                            NaN  

[5 rows x 228 columns]

In [7]:
#Veja a quantidade de linhas e de colunas do dataset
multiple_choice.shape

(16716, 228)

Existem 228 colunas!!!
![panda](https://media.giphy.com/media/14aUO0Mf7dWDXW/giphy.gif)

Vamos ver do que se tratam essas colunas. Como são MUITAS colunas precisamos alterar a configuração padrão do pandas para visualização de linhas e colunas

In [8]:
pd.set_option('max_rows', 200)
pd.set_option('max_columns', 1000)

In [9]:
# Vamos ver um pedaço do dataset
multiple_choice.head()

GenderSelect        Country   Age  \
0  Non-binary, genderqueer, or gender non-conforming            NaN   NaN   
1                                             Female  United States  30.0   
2                                               Male         Canada  28.0   
3                                               Male  United States  56.0   
4                                               Male         Taiwan  38.0   

                                    EmploymentStatus StudentStatus  \
0                                 Employed full-time           NaN   
1                 Not employed, but looking for work           NaN   
2                 Not employed, but looking for work           NaN   
3  Independent contractor, freelancer, or self-em...           NaN   
4                                 Employed full-time           NaN   

  LearningDataScience CodeWriter CareerSwitcher  \
0                 NaN        Yes            NaN   
1                 NaN        NaN            NaN   
2                 NaN        NaN            NaN   
3                 NaN        Yes            NaN   
4                 NaN        Yes            NaN   

              CurrentJobTitleSelect TitleFit  \
0             DBA/Database Engineer     Fine   
1                               NaN      NaN   
2                               NaN      NaN   
3  Operations Research Practitioner   Poorly   
4                Computer Scientist     Fine   

                                 CurrentEmployerType MLToolNextYearSelect  \
0  Employed by a company that doesn't perform adv...             SAS Base   
1                                                NaN               Python   
2                                                NaN  Amazon Web services   
3                                      Self-employed           TensorFlow   
4  Employed by a company that doesn't perform adv...           TensorFlow   

  MLMethodNextYearSelect LanguageRecommendationSelect  \
0         Random Forests                           F#   
1         Random Forests                       Python   
2          Deep learning                            R   
3            Neural Nets                       Python   
4            Text Mining                       Python   

                                PublicDatasetsSelect  \
0  Dataset aggregator/platform (i.e. Socrata/Kagg...   
1  Dataset aggregator/platform (i.e. Socrata/Kagg...   
2  Dataset aggregator/platform (i.e. Socrata/Kagg...   
3          I collect my own data (e.g. web-scraping)   
4                                             GitHub   

                              LearningPlatformSelect  \
0  College/University,Conferences,Podcasts,Trade ...   
1                                             Kaggle   
2  Arxiv,College/University,Kaggle,Online courses...   
3  Blogs,College/University,Conferences,Friends n...   
4                  Arxiv,Conferences,Kaggle,Textbook   

  LearningPlatformUsefulnessArxiv LearningPlatformUsefulnessBlogs  \
0                             NaN                             NaN   
1                             NaN                             NaN   
2                     Very useful                             NaN   
3                             NaN                     Very useful   
4                     Very useful                             NaN   

  LearningPlatformUsefulnessCollege LearningPlatformUsefulnessCompany  \
0                               NaN                               NaN   
1                               NaN                               NaN   
2                   Somewhat useful                               NaN   
3                       Very useful                               NaN   
4                               NaN                               NaN   

  LearningPlatformUsefulnessConferences LearningPlatformUsefulnessFriends  \
0                           Very useful                               NaN   
1                                   NaN                               Na

Podemos ver só o nome das colunas também utilizando o `columns`. Para ficar mais fácil de visualizar, ao invés de retornar o array, podemos transformar esse dado em uma Series.

In [10]:
# Use o columns no dataframe e coloque-o em uma Series para facilitar a visualização
pd.Series(multiple_choice.columns)

0                                     GenderSelect
1                                          Country
2                                              Age
3                                 EmploymentStatus
4                                    StudentStatus
5                              LearningDataScience
6                                       CodeWriter
7                                   CareerSwitcher
8                            CurrentJobTitleSelect
9                                         TitleFit
10                             CurrentEmployerType
11                            MLToolNextYearSelect
12                          MLMethodNextYearSelect
13                    LanguageRecommendationSelect
14                            PublicDatasetsSelect
15                          LearningPlatformSelect
16                 LearningPlatformUsefulnessArxiv
17                 LearningPlatformUsefulnessBlogs
18               LearningPlatformUsefulnessCollege
19               LearningPlatfo

Podemos ver mais detalhes do dataset com o `info()`

In [11]:
# use o info no dataframe
multiple_choice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16716 entries, 0 to 16715
Columns: 228 entries, GenderSelect to JobFactorPublishingOpportunity
dtypes: float64(13), object(215)
memory usage: 29.1+ MB


Podemos dar uma olhada nos tipos de campos que vem em cada uma das colunas númericas com um único comando

In [12]:
# veja as estatisticas dos campos númericos
multiple_choice.describe()

Age  LearningCategorySelftTaught  \
count  16385.000000                 13109.000000   
mean      32.372841                    33.366771   
std       10.473487                    25.787181   
min        0.000000                     0.000000   
25%       25.000000                    15.000000   
50%       30.000000                    30.000000   
75%       37.000000                    50.000000   
max      100.000000                   100.000000   

       LearningCategoryOnlineCourses  LearningCategoryWork  \
count                   13126.000000          13111.000000   
mean                       27.375514             15.217593   
std                        26.860840             18.996778   
min                         0.000000              0.000000   
25%                         5.000000              0.000000   
50%                        20.000000             10.000000   
75%                        40.000000             25.000000   
max                       100.000000            100.000000   

       LearningCategoryUniversity  LearningCategoryKaggle  \
count                13122.000000            13126.000000   
mean                    16.988607                5.531434   
std                     23.676917               11.072680   
min                      0.000000                0.000000   
25%                      0.000000                0.000000   
50%                      5.000000                0.000000   
75%                     30.000000               10.000000   
max                    100.000000              100.000000   

       LearningCategoryOther  TimeGatheringData  TimeModelBuilding  \
count           13094.000000        7530.000000        7528.000000   
mean                1.795940          36.144754          21.268066   
std                 9.357886          21.649591          16.165958   
min                 0.000000           0.000000           0.000000   
25%                 0.000000          20.000000          10.000000   
50%                 0.000000          35.000000          20.000000   
75%                 0.000000          50.000000          30.000000   
max               100.000000         100.000000         100.000000   

       TimeProduction  TimeVisualizing  TimeFindingInsights  TimeOtherSelect  
count     7517.000000      7529.000000          7523.000000      7513.000000  
mean        10.806372        13.869372            13.094776         2.396247  
std         12.257932        11.722945            12.974846        12.157137  
min          0.000000         0.000000             0.000000         0.000000  
25%          0.000000         5.000000             5.000000         0.000000  
50%         10.000000        10.000000            10.000000         0.000000  
75%         15.000000        20.000000            20.000000         0.000000  
max        100.000000       100.000000           303.000000       100.000000

E se eu quiser ver a quantidade de nulos no dataset todo?

In [13]:
# Veja nulos do dataset
multiple_choice.isnull().sum()

GenderSelect                                      95
Country                                          121
Age                                              331
EmploymentStatus                                   0
StudentStatus                                  15436
LearningDataScience                            15432
CodeWriter                                      3530
CareerSwitcher                                 13704
CurrentJobTitleSelect                           4886
TitleFit                                        5212
CurrentEmployerType                             5115
MLToolNextYearSelect                            5718
MLMethodNextYearSelect                          5883
LanguageRecommendationSelect                    5718
PublicDatasetsSelect                            5920
LearningPlatformSelect                          5445
LearningPlatformUsefulnessArxiv                14325
LearningPlatformUsefulnessBlogs                11951
LearningPlatformUsefulnessCollege             

E se eu quiser fazer a porcentagem de nulos?

In [14]:
# veja a porcentagem de nulos
multiple_choice.isnull().sum() / len(multiple_choice)

GenderSelect                                   0.005683
Country                                        0.007239
Age                                            0.019801
EmploymentStatus                               0.000000
StudentStatus                                  0.923427
LearningDataScience                            0.923187
CodeWriter                                     0.211175
CareerSwitcher                                 0.819813
CurrentJobTitleSelect                          0.292295
TitleFit                                       0.311797
CurrentEmployerType                            0.305994
MLToolNextYearSelect                           0.342067
MLMethodNextYearSelect                         0.351938
LanguageRecommendationSelect                   0.342067
PublicDatasetsSelect                           0.354152
LearningPlatformSelect                         0.325736
LearningPlatformUsefulnessArxiv                0.856963
LearningPlatformUsefulnessBlogs                0

Nossa quanto nulo!
![sad_panda](https://media.giphy.com/media/3e18NPUVzoxzO/giphy.gif)

O que eu devo fazer se eu quiser ver apenas o quanto o slário impacta para um cientista de dados? A coluna é `JobFactorSalary`

In [15]:
# veja uma única coluna
multiple_choice.JobFactorSalary
multiple_choice['JobFactorSalary']

0                       NaN
1                       NaN
2            Very Important
3                       NaN
4                       NaN
5                       NaN
6                       NaN
7            Very Important
8                       NaN
9                       NaN
10       Somewhat important
11                      NaN
12       Somewhat important
13                      NaN
14                      NaN
15                      NaN
16                      NaN
17                      NaN
18       Somewhat important
19            Not important
20       Somewhat important
21                      NaN
22                      NaN
23                      NaN
24                      NaN
25                      NaN
26                      NaN
27                      NaN
28                      NaN
29           Very Important
30                      NaN
31                      NaN
32                      NaN
33                      NaN
34                      NaN
35           Very Im

Vamos fazer algumas operações com o pandas para contar o número de nulos que existem nessa coluna

In [16]:
# Veja o número de nulos apenas desta coluna
multiple_choice.JobFactorSalary.isnull().sum()

13231

Como eu faço se eu só quiser ver os 10 primeiros registros?

In [17]:
# Selecione apenas os 10 primeiros registros
multiple_choice['JobFactorSalary'][:10]

0               NaN
1               NaN
2    Very Important
3               NaN
4               NaN
5               NaN
6               NaN
7    Very Important
8               NaN
9               NaN
Name: JobFactorSalary, dtype: object

E se eu quiser ver 2 colunas ao mesmo tempo? (E apenas essas 2 colunas)

In [18]:
# Veja as colunas 'JobFactorSalary' e 'JobFactorLearning'
multiple_choice[['JobFactorSalary', 'JobFactorLearning']][:10]

JobFactorSalary JobFactorLearning
0             NaN               NaN
1             NaN               NaN
2  Very Important    Very Important
3             NaN               NaN
4             NaN               NaN
5             NaN               NaN
6             NaN               NaN
7  Very Important    Very Important
8             NaN               NaN
9             NaN               NaN

O quanto que as pessoas dessa pesquisa estão satisfeitas com o trabalhos? Conseguimos saber isso usando só o pandas?

In [19]:
# Use uma função para contar o número de cada uma das variáveis da coluna JobSatisfaction
multiple_choice['JobSatisfaction'].value_counts()

7                          1448
8                          1427
6                           765
9                           677
5                           627
10 - Highly Satisfied       589
3                           358
4                           354
1 - Highly Dissatisfied     167
I prefer not to share       148
2                           117
Name: JobSatisfaction, dtype: int64

Percebemos com esse comando que as pessoas até que estão bastante satisfeitas.

Agora vamos olhar só as pessoas que estão Super Satisfeitas (Highly Satisfied) com o seu trabalho. Como que eu posso fazer isso?

In [20]:
# Filtre só quem está com o JobSatisfaction de 10. Guarde isso em uma variável pq é bastante dado
highly_satisfied = multiple_choice[multiple_choice['JobSatisfaction'] == '10 - Highly Satisfied']
highly_satisfied

GenderSelect         Country  \
3                                                   Male   United States   
60                                                  Male          Canada   
77                                                  Male          Israel   
83                                                  Male          Russia   
89                                                Female   United States   
95                                                  Male          Canada   
100                                               Female   United States   
127                                                 Male          Brazil   
130                                                 Male          Brazil   
131                                                 Male           India   
134                                                 Male           Other   
176                                                 Male   United States   
179                                                 Male          France   
183                                                 Male           India   
200                                                 Male   United States   
203                                                 Male   United States   
217                                                 Male   United States   
283                                                 Male   United States   
316                                                 Male           Japan   
348                                               Female   United States   
360                                                 Male           Other   
368                                                 Male          France   
393                                                 Male           Chile   
419                                                 Male   United States   
425                                               Female   United States   
429                                                 Male          France   
434                                                 Male   United States   
448                                                 Male   United States   
461                                                 Male   United States   
473                                                 Male       Argentina   
486                                                 Male           India   
516                                                 Male   United States   
518                                               Female   United States   
531                                                 Male   United States   
545                                                 Male          Russia   
566                                                 Male         Germany   
575                                                 Male          Canada   
594                                                 Male   United States   
595                                               Female   United States   
600                                               Female   United States   
633                                                 Male          Canada   
650                                                 Male          Canada   
687                                                 Male          Canada   
710                                                 Male           Japan   
731                                                 Male   United States   
808                                               Female   United States   
814                                                 Male   United States   
852                                                 Male   United States   
912                                               Female   United States   
917                                               Female   United States   
954                                                 Male           India   
957                                                 Male       Argentina   
975             

In [21]:
# Veja o tamanho do dataset. Ele bateu com a quantidade de pessoas que estão altamente satisfeitas?
highly_satisfied.shape

(589, 228)

In [22]:
# Veja os primeiros 3 registros (todas as colunas) das pessoas altamentes satisfeitas
highly_satisfied[:3]

GenderSelect        Country   Age  \
3          Male  United States  56.0   
60         Male         Canada  34.0   
77         Male         Israel  29.0   

                                     EmploymentStatus StudentStatus  \
3   Independent contractor, freelancer, or self-em...           NaN   
60  Independent contractor, freelancer, or self-em...           NaN   
77  Independent contractor, freelancer, or self-em...           NaN   

   LearningDataScience CodeWriter CareerSwitcher  \
3                  NaN        Yes            NaN   
60                 NaN        Yes            NaN   
77                 NaN        Yes            NaN   

               CurrentJobTitleSelect   TitleFit CurrentEmployerType  \
3   Operations Research Practitioner     Poorly       Self-employed   
60                             Other       Fine       Self-employed   
77                    Data Scientist  Perfectly       Self-employed   

   MLToolNextYearSelect MLMethodNextYearSelect LanguageRecommendationSelect  \
3            TensorFlow            Neural Nets                       Python   
60                  NaN                    NaN                       Python   
77                  NaN          Deep learning                       Python   

                                 PublicDatasetsSelect  \
3           I collect my own data (e.g. web-scraping)   
60                                                NaN   
77  Dataset aggregator/platform (i.e. Socrata/Kagg...   

                               LearningPlatformSelect  \
3   Blogs,College/University,Conferences,Friends n...   
60  Arxiv,Blogs,Kaggle,Online courses,Personal Pro...   
77  Arxiv,Kaggle,Official documentation,Online cou...   

   LearningPlatformUsefulnessArxiv LearningPlatformUsefulnessBlogs  \
3                              NaN                     Very useful   
60                             NaN                             NaN   
77                 Somewhat useful                             NaN   

   LearningPlatformUsefulnessCollege LearningPlatformUsefulnessCompany  \
3                        Very useful                               NaN   
60                               NaN                               NaN   
77                               NaN                               NaN   

   LearningPlatformUsefulnessConferences LearningPlatformUsefulnessFriends  \
3                            Very useful                       Very useful   
60                                   NaN                               NaN   
77                                   NaN                               NaN   

   LearningPlatformUsefulnessKaggle LearningPlatformUsefulnessNewsletters  \
3                               NaN                                   NaN   
60                              NaN                                   NaN   
77                      Very useful                                   NaN   

   LearningPlatformUsefulnessCommunities  \
3                                    NaN   
60                                   NaN   
77                                   NaN   

   LearningPlatformUsefulnessDocumentation LearningPlatformUsefulnessCourses  \
3                              Very useful                       Very useful   
60                                     NaN                               NaN   
77                             Very useful                       Very useful   

   LearningPlatformUsefulnessProjects LearningPlatformUsefulnessPodcasts  \
3                         Very useful                                NaN   
60                                NaN                                NaN   
77                                NaN                                NaN   

   LearningPlatformUsefulnessSO LearningPlatformUsefulnessTextbook  \
3                           NaN                                NaN   
60                          NaN                                NaN   
77                  Very useful                                NaN   

   L

E se eu quiser ver as pessoas altamente satisfeitas e que trabalham com python?

In [23]:
# Preciso usar 2 filtros. JobSatisfaction e LanguageRecommendationSelect
highly_satisfied = multiple_choice['JobSatisfaction'] == '10 - Highly Satisfied'
pythonist = multiple_choice['LanguageRecommendationSelect'] == 'Python'
highly_satisfied_and_pythonist = multiple_choice[highly_satisfied & pythonist]

In [24]:
# Só para validar que deu certo é bom ver o shape do resultado
highly_satisfied_and_pythonist.shape

(319, 228)

E se tentassemos com a idade? Ver só que está abaixo de 30 anos

In [25]:
# Aqui precisa usar JobSatisfaction e Age
multiple_choice['Age'].value_counts()
age = multiple_choice['Age'] < 30.0
highly_satisfied_and_age = multiple_choice[highly_satisfied & age]

In [26]:
# Só para validar que deu certo é bom ver o shape do resultado
highly_satisfied_and_age.shape

(171, 228)

Quais são as linguagens que a galera altamente satisfeita recomenda? Como vc faria para mostrar esses valores em porcentagem?

In [27]:
# preciso usar a variável que eu criei umas células acima e LanguageRecommendationSelect e também do número de pessoas que responderam a pergunta JobSatisfaction
highly_satisfied
highly_languages = multiple_choice[highly_satisfied]['LanguageRecommendationSelect']
highly_languages

3          Python
60         Python
77         Python
83              R
89              R
95         Python
100        Python
127        Python
130        Python
131        Python
134        Python
176         Julia
179             R
183        Python
200        Python
203        Python
217        Python
283        Python
316        Python
348        Python
360        Python
368        Python
393        Python
419        Python
425        Python
429             R
434        Python
448        Python
461        Python
473        Python
486             R
516        Python
518             R
531             R
545        Python
566        Python
575             R
594         Other
595           SQL
600           SQL
633           SQL
650        Python
687        Python
710        Python
731        Python
808             R
814        Python
852        Python
912      C/C++/C#
917             R
954             R
957             R
975        Python
1003       Python
1024            R
1033      

E se eu quiser ordenar esses valores? Do menor para o maior?

In [32]:
language_counts = highly_languages.notnull().sum()
(highly_languages.value_counts() / len(highly_languages) * 100).sort_values()

Haskell      0.339559
Stata        0.509338
SAS          0.679117
Julia        0.848896
Java         1.188455
Scala        1.528014
Other        1.528014
Matlab       1.867572
C/C++/C#     3.395586
SQL          4.753820
R           25.636672
Python      54.159593
Name: LanguageRecommendationSelect, dtype: float64

### Desafio 1

Qual a coluna que tem o menor número de dados? Com essa informação em mãos, conseguimos dizer qual o país que mais tem registros nessa coluna.

Dica: Você precisará ordenar os campos pela quantidade de nulos (ou não nulos) e depois ver o país dessa galera.

![arrested_panda](https://media.giphy.com/media/N6funLtVsHW0g/giphy.gif)

In [56]:
# primeiro descubra a coluna que tem maior número de campos nulos
null_colunm = multiple_choice.isnull().sum().sort_values(ascending=False)
null_colunm
# depois extraia o pais que mais tem respostas nessa coluna
multiple_choice[multiple_choice['WorkToolsFrequencyAngoss'].notnull()]['Country'].value_counts()

United States     8
India             3
Italy             2
Canada            2
Mexico            2
Other             1
Egypt             1
Australia         1
United Kingdom    1
Singapore         1
Name: Country, dtype: int64